### PR TITLE
Mypy: Error on Common Truthy Mistakes

### PR DESCRIPTION
**Pull request by MattToast**
Configure mypy to raise an error when:
- An instance of an object is used for a boolean check when neither `__bool__` or `__len__` are implemented
- A `Iterator` is used on a boolean check when the author almost certainly wanted a `Collection`

Fixes up/refactors areas of the code base where these potential errors linger.

### CODE REVIEW

File: smartsim/entity/ensemble.py [hunk 1]
```diff
@@ -104,12 +104,9 @@ def __init__(
...
 
     @property
-    def models(self) -> t.Iterable[Model]:
-        """
-        Helper property to cast self.entities to Model type for type correctness
-        """
-        model_entities = [node for node in self.entities if isinstance(node, Model)]
-        return model_entities
+    def models(self) -> t.Collection[Model]:
+        """An allias for a shallow copy of the ``entities`` attribute"""
```
Comment (ankona) [inline smartsim/entity/ensemble.py:108]: alias!

File: doc/changelog.rst [hunk 2]
```diff
@@ -48,7 +48,11 @@ Detailed Notes
   software and files that come from the default Github Ubuntu container.
   (SmartSim-PR504_)
 - Update the generic `t.Any` typehints in Experiment API. (SmartSim-PR501_)
+- The CI will fail static analysis if common erroneous tuthy checks are
```
Comment (al-rigazzi) [inline doc/changelog.rst:51]: `tuthy` -> `truthy`?

**Review by al-rigazzi — APPROVED** [review]
Just a minor typo in the changelog, otherwise LGTM!

File: smartsim/entity/ensemble.py (hunk 1 above)
Comment (MattToast) [reply smartsim/entity/ensemble.py:108]: 🤦

File: doc/changelog.rst (hunk 2 above)
Comment (MattToast) [reply doc/changelog.rst:51]: spelling is hard :(

File: smartsim/entity/model.py [hunk 3]
```diff
@@ -428,40 +428,34 @@ def _stringify_id(_id: int) -> str:
 
             raise TypeError(f"Argument is of type '{type(_id)}' not 'int'")
 
-        _invalid_input_message = (
-            "Expected a cpu pinning specification of type iterable of ints or "
-            f"iterables of ints. Instead got type `{type(pin_ids)}`"
-        )
+        try:
```
Comment (ankona) [inline smartsim/entity/model.py:431]: While you're in here, could you tweak/fix the docstring? My inner pedant would be stoked if you swap `Create a comma-separated string CPU ids` with `Create a comma-separated string of CPU ids` and wrap None w/backticks `None`

**Review by ankona — APPROVED** [review]
LGTM

File: smartsim/entity/model.py (hunk 3 above)
Comment (MattToast) [reply smartsim/entity/model.py:431]: Totally reasonable! Will add!!

**Review by al-rigazzi — COMMENTED** [review]
Still LGTM

**Pull request merged**